### DISCUSSION TITLE
MML support

### DISCUSSION THREAD

**Discussion by soywod** [stalwartlabs/mail-builder#7] [Ideas]
Hi there,

I just discovered all your mail related projects (thanks to [nlnet](https://nlnet.nl/project/Stalwart/index.html)), congrats for your giant work!

I am also working an an mail related project called [Himalaya](https://github.com/soywod/himalaya) (also helped by [nlnet](https://nlnet.nl/project/Himalaya/index.html)). It was initially a CLI to manage mails, but it became over time an abstract [Rust lib](https://git.sr.ht/~soywod/himalaya-lib) so you can build any kind of interface on the top of it (the aim is not to reimplement common actions like read an email, copy/move/delete an email, add a flag etc and to be backend agnostic: IMAP, POP, Maildir, Notmuch etc). The CLI became the first interface based on this lib. So far I also have a [Vim plugin](https://git.sr.ht/~soywod/himalaya-vim) and an [Emacs plugin](https://github.com/dantecatalfamo/himalaya-emacs) based on the CLI.

I contact you because I developed a [MIME Message builder](https://git.sr.ht/~soywod/mime-msg-builder) which achieves the same thing as your `mail-builder` (it is just a wrapper around the [lettre](https://github.com/lettre/lettre) message builder), except that it can also interpret [MML](https://www.gnu.org/software/emacs/manual/html_node/emacs-mime/Composing.html). MML is a DSL coming from an Emacs module allowing you to compile template (string) into valid MIME messages. From their [doc](https://www.gnu.org/software/emacs/manual/html_node/emacs-mime/Simple-MML-Example.html):

```
<#multipart type=alternative>
This is a plain text part.
<#part type=text/enriched>
<center>This is a centered enriched part</center>
<#/multipart>
```

becomes:

```
Content-Type: multipart/alternative; boundary="=-=-="


--=-=-=


This is a plain text part.

--=-=-=
Content-Type: text/enriched


<center>This is a centered enriched part</center>

--=-=-=--
```

It is necessary for me, because it allows clients to build valid MIME messages. They can also add multiple parts, add attachments etc, directly from the message edition. It is very handy.

My question is: would you be interested in adding such a feature to your crate? You already have `serde` as a dependency, and the MML can be seen as a string representation of a message. Here some ideas of usage:

```rust
// feed an existing message builder with MML
let eml = MessageBuilder::new()
    .from(("John Doe", "john@doe.com"))
    .to("jane@doe.com")
    .subject("Hello, world!")
    .mml("<#part type=image/jpeg filename=~/rms.jpg disposition=inline>");

// export a message builder to MML
let mml_string = MessageBuilder::new()
    .from(("John Doe", "john@doe.com"))
    .to("jane@doe.com")
    .subject("Hello, world!")
    .write_to_mml_string();

// parse a string into a message builder then override headers
let mut eml: MessageBuilder = "From: John Doe <john@doe.com>
To: jane@doe.com
Subject: Hello, world!

<#part type=image/jpeg filename=~/rms.jpg disposition=inline>".parse().unwrap();
eml = eml
    .to("jane-override@doe.com")
    .subject("Hello, world 2!");
```

While writing this message, I thought that maybe a better place for this feature would be in your `mail-parser` crate. But for sure, nice integration can be done between crates.

Let me know!

**Comment by mdecimus** [comment]
Hi @soywod 

Thanks for the suggestion! 

I checked MML and,  unless it becomes a IETF standard to replace MIME, I think it is beyond the scope of both `mail-parser` and `mail-builder`. The goal of `mail-builder` is to programmatically generate valid MIME messages from Rust and, in my opinion, MML just adds more complexity to that process by requiring users to learn a new DSL. In addition to that there is a performance penalty of having to parse the MML DSL (although the parsing could be implemented as a macro). In the case of `mail-parser`, the goal of the library is to parse messages sent using any of the IETF standardized formats for transmitting internet messages (RFC5322, MIME, etc) and MML does not fit this description as it is a templating language that gets converted into MIME.
That being said, MML does look clearer and less verbose than MIME and I hope it replaces MIME one day, but until then, I think it makes more sense to have a separate crate to handle MML processing.

BTW, I wasn't aware of Himalaya, I will definitely check it out once I start working on the new version of the IMAP server!

Cheers!

**Reply by soywod** [reply]
Thanks for your quick reply, and I totally understand. I just proposed because I see that both of our crates are close but not close enough to be together. I think I will just develop a wrapper around your builder to support MML.

Do you plan to develop an IMAP client by any chance? For now I use [rust-imap](https://github.com/jonhoo/rust-imap), it works well but it looks like I have perfs issues with. I ask this question because I saw you developed a JMAP client and I will definitely use it for [#35](https://todo.sr.ht/~soywod/pimalaya/35). Nice to see that we have complementary projects :) and nice to meet you! See y'a.

PS: we have a [matrix room](https://matrix.to/#/#pimalaya:matrix.org), feel free to join if it interests you!

**Reply by mdecimus** [reply]
> Thanks for your quick reply, and I totally understand. I just proposed because I see that both of our crates are close but not close enough to be together. I think I will just develop a wrapper around your builder to support MML.

Sounds good!

> Do you plan to develop an IMAP client by any chance? For now I use [rust-imap](https://github.com/jonhoo/rust-imap), it works well but it looks like I have perfs issues with. 

Not really, my focus is on having a robust mail server written in Rust.

> I ask this question because I saw you developed a JMAP client and I will definitely use it for [#35](https://todo.sr.ht/~soywod/pimalaya/35). Nice to see that we have complementary projects :) and nice to meet you! See y'a.

Nice to meet you too!
